### ISSUE TITLE
Potential Misstatement Regarding LayerNorm in Section 12.4

### ISSUE THREAD

**Issue by Vanster78** [udlbook/udlbook#237]
Hi Simon,

First of all, thank you so much for your incredible work. Reading this book has been a great pleasure for me, and I've learned a lot.

I would like to discuss some potential misstatements I've noticed.

In section 12.4 there's a statement:

> In addition, it is typical to add a LayerNorm operation after both the self-attention and fully connected networks. This is similar to BatchNorm but uses statistics across the tokens within a single input sequence to perform the normalization (section 11.4 and figure 11.14).

If I understand correctly, it seems to suggest that the means and variances are computed for the entire sequence and shared between all tokens. I believe this is not the case. From my understanding, LayerNorm in transformers is applied individually to each token, meaning it centers and standardizes each token’s embedding separately, and then scales and shifts its components with shared learned parameters.

I agree that it's quite misleading and is rather similar to InstanceNorm since it doesn't leverage spatial information, but here are some thoughts to support this: 

<h4>1. Implementation references</h4>

In common transformer implementations (e.g. Andrew Karpathy's [minGPT](https://github.com/karpathy/minGPT/blob/master/mingpt/model.py)) LayerNorm is initialized like this:

`self.ln_1 = nn.LayerNorm(config.n_embd)`

So, according to the [PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html), it just gets embedding size as _normalized_shape_:

> If a single integer is used, it is treated as a singleton list, and this module will normalize over the last dimension which is expected to be of that specific size.

Thus, for a tensor of shape [B, N, C], such a module operates along the last dimension, i.e. the channels of each individual token.

<h4>2. Data leakage</h4>

In transformer decoders, information is only shared across tokens within masked self-attention. If LayerNorm were applied across tokens, this would cause data leakage from future tokens, which would be inappropriate for casual language modeling.

<h3>Proposed reformulation</h3>

> In addition, it is typical to add a LayerNorm operation after both the self-attention and fully connected networks. This is similar to BatchNorm but performs the normalization individually for each token embedding (section 11.4 and figure 11.14).

<br/>

Thanks again and warm regards,
Ivan

**Comment by udlbook** [comment]
Hi Ivan, 

Thank you so much for taking the time to message me and provide this detailed argument.  I actually didn't know about this, but I've done some Googling and you are correct.  The data leakage argument is pretty convincing anyway.  

I'll add your name to the acknowledgements and make the change.  In fact, I think I should change Figure 12.7 as well, so that the LayerNorm is shown operating separately on each token.
<img width="908" alt="Screenshot 2024-10-15 at 5 56 09 PM" src="https://github.com/user-attachments/assets/5eb2d537-3c95-4ab5-88c1-0e8d4f34d990">

**Comment by Vanster78** [comment]
Yes, the picture now looks more valid to me! Thank you for taking the time to consider it.
